### PR TITLE
Oo implementation

### PR DESCRIPTION
**Pull request by thoeschler**

### CODE REVIEW

File: Solver/elastic_law.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+import sympy as sp
+import numpy as np
+import itertools as it
+from shenfun import inner, div, grad, Dx
+
+
+class LinearCauchyElasticity:
+    def __init__(self):
+        self.name = "LinearCauchyElasticity"
+        self.nb_material_parameters = 2
```
Comment (sebglane) [inline Solver/elastic_law.py]: I prefer `n_material_parameters` if this variable refers to the number of material parameters.

Comment (sebglane) [inline Solver/elastic_law.py]: You could also make these variables private by adding an underscore, i.e., `self._name`.

File: Solver/elastic_law.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+        error = np.sqrt(inner((1, 1), error_array ** 2))
+        # scale by magnitude of solution
+        scale = np.sqrt(inner((1, 1), u_hat.backward() ** 2))
+
+        return error / scale
+
+    def compute_body_forces(self, u):
+        assert hasattr(self, "material_parameters")
+        assert len(self.material_parameters) == self.nb_material_parameters
```
Comment (sebglane) [inline Solver/elastic_law.py]: Shouldn't there be an assertion to guarantee that `u` is a `sympy.Expr`?

File: Solver/elastic_law.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+
+class LinearCauchyElasticity:
+    def __init__(self):
+        self.name = "LinearCauchyElasticity"
+        self.nb_material_parameters = 2
+
+    def check_with_pde(self, u_hat, material_parameters, body_forces):
+        assert isinstance(u_hat, sf.Function)
+        assert len(material_parameters) == self.nb_material_parameters
```
Comment (sebglane) [inline Solver/elastic_law.py]: Shouldn't there be an assertion to guarantee that `body_forces` has a compatible type?

File: Solver/elastic_law.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+        body_forces = - (lmbda + mu) * GradDiv - mu * Laplace
+
+        return tuple(body_forces)
+
+    def compute_cauchy_stresses(self, u_hat):
+        assert isinstance(u_hat, sf.Function)
+
+        space = u_hat[0].function_space().get_orthogonal()
+        dim = len(space.bases)
+        N = [u_hat.function_space().spaces[0].bases[i].N for i in range(dim)]
```
Comment (sebglane) [inline Solver/elastic_law.py:66]: What is the meaning of `N` here? Number of degrees of freedom in each component? Please add a comment.

File: Solver/elastic_law.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        # linear strain tensor
+        E = 0.5 * (H + np.swapaxes(H, 0, 1))
+
+        # trace of linear strain tensor
+        trE = np.trace(E)
+
+        # create block with identity matrices on diagonal
+        identity = np.zeros_like(H)
+        for i in range(dim):
+            identity[i, i] = np.ones(N)
```
Comment (sebglane) [inline Solver/elastic_law.py:84]: `np.ones` return a vector of ones with length `N`. Of course, this is equivalent to an identity matrix but then `identity` contains vectors whereas `E` is a `np.ndarray` of, I guess, `shenfun.Array`.

File: Solver/elastic_law.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+        return dw_int
+
+    def set_material_parameters(self, material_parameters):
+        self.material_parameters = material_parameters
+
+
+class LinearGradientElasticity:
+    def __init__(self):
+        self.name = "LinearGradientElasticity"
+        self.nb_material_parameters = 7
```
Comment (sebglane) [inline Solver/elastic_law.py]: Same here.

File: Solver/elastic_law.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+
+    def check_with_pde(self, u_hat, material_parameters, body_forces):
+        assert isinstance(u_hat, sf.Function)
+        assert len(material_parameters) == self.nb_material_parameters
```
Comment (sebglane) [inline Solver/elastic_law.py]: Same here regarding the type `body_forces`.

File: Solver/elastic_law.py [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+
+        lambd, mu, c1, c2, c3, c4, c5 = material_parameters
+        V = u_hat.function_space().get_orthogonal()
+        # left hand side of pde
+        lhs = (lambd + mu) * grad(div(u_hat)) + mu * div(grad(u_hat)) \
+            - (c1 + c4) * div(grad(div(grad(u_hat)))) \
+            - (c2 + c3 + c5) * grad(div(div(grad(u_hat))))
```
Comment (sebglane) [inline Solver/elastic_law.py:144]: Does this also work in 3D? If not you should catch this by an assertion.

File: Solver/elastic_law.py [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+        # Cauchy stress tensor
+        T = 2.0 * mu * E + lmbda * trE * identity
+
+        return T, space
+    
+    def compute_hyper_stresses(self, u_hat):
+        assert isinstance(u_hat, sf.Function)
+
+        space = u_hat[0].function_space()
+        dim = len(space.bases)
```
Comment (sebglane) [inline Solver/elastic_law.py:190]: Same here regarding the spatial dimension.

File: Solver/elastic_law.py [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+                            Dx(Dx(u_hat[i], j), k), space
+                            ).backward()
+
+        identity = np.identity(dim)
+
+        # hyper stresses
+        T = c1 * np.swapaxes(
+                np.tensordot(identity, Laplace, axes=0), 0, 2
+                ) \
+            + c2 / 2 * (np.tensordot(identity, Laplace, axes=0) +
```
Comment (sebglane) [inline Solver/elastic_law.py]: `c2 / 2.0`!

File: Solver/elastic_law.py [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+                        np.swapaxes(np.tensordot(identity, Laplace, axes=0),
+                                     1, 2)
+                        ) \
+            + c3 / 2 * (np.tensordot(identity, GradDiv, axes=0) +
```
Comment (sebglane) [inline Solver/elastic_law.py]: Same here.

File: Solver/elastic_law.py [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+                        np.swapaxes(np.tensordot(identity, GradDiv, axes=0),
+                                    1, 2)
+                        ) \
+            + c4 * GradGrad \
+            + c5 / 2 * (np.swapaxes(GradGrad, 0, 1) +
```
Comment (sebglane) [inline Solver/elastic_law.py]: and here.

File: Solver/elastic_law.py [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+            for i in range(self.dim):
+                for j in range(self.dim):
+                    for k in range(self.dim):
+                        mat = inner(c5 * Dx(Dx(u[j], i), k),
+                                    Dx(Dx(v[i], j), k))
+                        if isinstance(mat, list):
+                            dw_int += mat
+                        else:
+                            dw_int += [mat]
+        dw_int += inner(mu * grad(u), grad(v))
```
Comment (sebglane) [inline Solver/elastic_law.py:267]: Add a comment that the "classical terms" are added next.

File: Solver/elastic_law.py [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+                if isinstance(mat, list):
+                    dw_int += mat
+                else:
+                    dw_int += [mat]
+        dw_int += inner(lmbda * div(u), div(v))
+
+        return dw_int
+
+    def compute_body_forces(self, u):
+        assert isinstance(u, tuple)
```
Comment (sebglane) [inline Solver/elastic_law.py]: You should also check the type of the tuple entries.

File: Solver/elastic_law.py [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+        assert hasattr(self, "material_parameters")
+        assert len(self.material_parameters) == self.nb_material_parameters
+
+        dim = len(u)
```
Comment (sebglane) [inline Solver/elastic_law.py:286]: Same here regarding the spatial dimension.

File: Solver/elastic_law.py [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+            GradDivDivGrad[i] = DivDivGrad.diff(coord[i])
+
+        body_forces = (c1 + c4) * DoubleLaplace + \
+            (c2 + c3 + c5) * GradDivDivGrad - \
+            (lmbda + mu) * GradDiv - mu * Laplace
+
+        return tuple(body_forces)
+
+    def set_material_parameters(self, material_parameters):
+        self.material_parameters = material_parameters
```
Comment (sebglane) [inline Solver/elastic_law.py]: Here `material_parameters` should become a property with a setter and getter method. By the way, there is no consistency check regarding the number of material parameters and the type of the value.

File: Solver/elastic_problem.py [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+            self.l_ref = self.domain[0][1]
+            self.u_ref = 1
+            self.mat_param_ref = self.material_parameters[0]
+
+        # dl means dimensionless
+        self.domain_dl, self.bc_dl, self.body_forces_dl, \
+            self.material_parameters_dl = get_dimensionless_parameters(
+                    self.domain, self.bc, self.body_forces,
+                    self.material_parameters, self.u_ref, self.l_ref,
+                    self.mat_param_ref)
```
Comment (sebglane) [inline Solver/elastic_problem.py]: Maybe `self._dimless_domain` etc. is better...

File: Solver/utilities.py [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+         for i in range(dim)]
+
+    domain_dimless = [u_hat_dimless.function_space().spaces[i].bases[i].domain
+                      for i in range(dim)]
+    domain = [
+            [
+                domain_dimless[i][j] * l_ref for j in range(2)
+                ]
+            for i in range(dim)
+        ]
```
Comment (sebglane) [inline Solver/utilities.py]: Ugly indentation...

File: Solver/utilities.py [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+
+
+def get_dimensionless_material_parameters(material_parameters,
+                                          mat_param_ref, l_ref):
+    if len(material_parameters) == 2:  # cauchy elasticity
+        return tuple(np.array(material_parameters) / mat_param_ref)
+    elif len(material_parameters) == 7:  # gradient elasticity
+        lame_parameters = np.array(
+                material_parameters
+                )[[0, 1]] / mat_param_ref
```
Comment (sebglane) [inline Solver/utilities.py]: Ugly indentation...

File: tests/test_solver.py [hunk 20]
```diff
@@ -0,0 +1,487 @@
...
+                    ((0, self.u0), None),
+                    ((0, 0), None)
+                    )
+        elif self.elastic_law.name == "LinearGradientElasticity":
+            bc = ((
+                    {'left': [('D', 0.), ('N', 0.)],
+                        'right': [('D', self.u0)]},
+                    None),
+                  ((0., 0.), None)
+                  )
```
Comment (sebglane) [inline tests/test_solver.py]: Ugly indentation...

File: Solver/utilities.py [hunk 21]
```diff
@@ -0,0 +1,202 @@
...
+    # assign bc type to each boundary condition
+    def assign_bc_type(bc):
+        if bc is None:
+            return 'ZERO_TRACTION'
+        elif isinstance(bc, tuple):  # dirichlet-bcs are given as tuple
+            return 'DIRICHLET'
+        elif isinstance(bc, dict):  # all bcs can be defined via a dictionary
+            return 'ARBITRARY'
+        else:
+            raise NotImplementedError()
```
Comment (sebglane) [inline Solver/utilities.py:68]: Should we use an `enum` to handle the different boundary conditions? This might be a better than using strings.

**Review by sebglane — CHANGES_REQUESTED** [review]
Overall great work. Only minor remarks but quite a lot...

I need to check

- Solver/solver.py
- Solver/utilities.py
- tests/test_solver.py

with greater detail. Especially regarding the boundary conditions.

File: Solver/elastic_law.py (hunk 5 above)
Comment (thoeschler) [reply Solver/elastic_law.py:84]: `N` is a tuple or list like `(20, 20)` so this should produce a matrix conatining all ones which has the same shape as the corresponding `shenfun.Array`.

File: Solver/utilities.py (hunk 21 above)
Comment (thoeschler) [reply Solver/utilities.py:68]: I agree. I will implement this in the next commit.

Comment (sebglane) [reply Solver/utilities.py:68]: Do you mean on a separate branch?

File: Solver/elastic_law.py (hunk 5 above)
Comment (sebglane) [reply Solver/elastic_law.py:84]: Thank you for clarifying this.

**Comment by sebglane** [comment]
The only thing which needs to be fixed, is my comment regarding the spatial dimension. If it only works in 2D for now I suggest adding
```python3
assert dim == 2
```

File: Solver/utilities.py (hunk 21 above)
Comment (thoeschler) [reply Solver/utilities.py:68]: Yes, I think it would be better to implement this on a separate branch along with the traction bcs.

**Comment by thoeschler** [comment]
> The only thing which needs to be fixed, is my comment regarding the spatial dimension. If it only works in 2D for now I suggest adding
> 
> ```python
> assert dim == 2
> ```

I will add this to the `__init__` of the `ElasticSolver`.

**Review by thoeschler — COMMENTED** [review]
The last commit essentially just makes most variables private that should be private. The test file is now a bit more compact.

**Review by sebglane — APPROVED** [review]
